scripts/build-optimize-fs-size.py

```diff
@@ -0,0 +1,253 @@
+
+    def hardlink(self, src):
+        if DRY_RUN:
+           print(f'hardlink {self.path} {src}')
+           return
+        st = self.stats
+        os.remove(self.path)
+        os.link(src.path, self.path)
+        os.chmod(self.path, st.st_mode)
+        os.chown(self.path, st.st_uid, st.st_gid)
+        os.utime(self.path, times=(st.st_atime, st.st_mtime))
+
+    @property
+    def name(self):
+        return os.path.basename(self.path)
+
+    @cached_property
+    def stats(self):
+        return os.stat(self.path)
+
+    @cached_property
+    def size(self):
+        return self.stats.st_size
+
+    @cached_property
+    def checksum(self):
+        with open(self.path, 'rb') as f:
+            return hashlib.md5(f.read()).hexdigest()
+
+class FileManager:
+    def __init__(self, path):
+        self.path = path
+        self.files = []
+        self.folders = []
+        self.nindex = defaultdict(list)
+        self.cindex = defaultdict(list)
+
+    def add_file(self, path):
+        if not os.path.isfile(path) or os.path.islink(path):
+            return
+        f = File(path)
+        self.files.append(f)
+
+    def load_tree(self):
+        self.files = []
+        self.folders = []
+        for root, _, files in os.walk(self.path):
+            self.folders.append(File(root))
+            for f in files:
+                self.add_file(os.path.join(root, f))
+        print(f'loaded {len(self.files)} files and {len(self.folders)} folders')
+
+    def generate_index(self):
+        print('Computing file hashes')
+        for f in self.files:
+            self.nindex[f.name].append(f)
+            self.cindex[(f.name, f.checksum)].append(f)
+
+    def create_hardlinks(self):
+        print('Creating hard links')
+        for files in self.cindex.values():
+            if len(files) <= 1:
+                continue
+            orig = files[0]
+            for f in files[1:]:
+                f.hardlink(orig)
+
+class FsRoot:
+    def __init__(self, path):
+        self.path = path
+
+    def iter_fsroots(self):
+        yield self.path
+        dimgpath = os.path.join(self.path, 'var/lib/docker/overlay2')
+        for layer in os.listdir(dimgpath):
+            yield os.path.join(dimgpath, layer, 'diff')
+
+    def collect_fsroot_size(self):
+        cmd = ['du', '-sb', self.path]
+        p = subprocess.run(cmd, text=True, check=False,
+                           stdout=subprocess.PIPE, stderr=subprocess.DEVNULL)
+        return int(p.stdout.split()[0])
+
+    def _remove_root_paths(self, relpaths):
+        for root in self.iter_fsroots():
+            for relpath in relpaths:
+                path = os.path.join(root, relpath)
+                if os.path.isdir(path):
+                    if DRY_RUN:
+                        print(f'rmtree {path}')
+                    else:
+                        shutil.rmtree(path)
+
+    def remove_docs(self):
+        self._remove_root_paths([
+            'usr/share/doc',
```

Comment:
Some of these cleanup, especially the removal ones could definitely be added as part of the general build.

However this would have to be more than just here, since it would also need to apply to all docker images.
On top of that the removal of licenses could have some legal implication for public/redistributable builds such as the community ones.
This should however not be a problem for internal/private images.

Most of the savings of this change actually come from the hardlinking steps.
And for Aboot images the removal of other platforms and unneeded kernel modules and firmwares.

This mitigation, even if it merges in master, has 202305 as the main target for 4GB systems.
It therefore felt simpler to centralize the cleanup under a postprocessing script hidden behind a toggle instead of touching tens of files.
This script also has the advantage of cleaning up behind newly introduced code that forgot the cleanup steps.
Until there is some checks and tooling in place to prevent new code from contributing to bloat, things will slip and become a maintenance burden.

Reply:
To add to your comment, the proper thing to do here would be to add an `/etc/dpkg/dpkg.conf.d/01_sonic` conf  instead of removing files like it's currently done in build_debian.sh
In this configuration you can specify things like `path-exclude=/usr/share/doc` however that will not retroactively apply the previously installed packages.

In the case of docker, that means that the debian base image will retain these files no matter what.
If you decide to remove them as part of the build of a child Dockerfile, this is actually going to create whiteout files which consumes space instead of the final goal being to release space.